File: changji/src/changji/assembly/subtitles.py
"""中文字幕生成。

用 ASS 不用 SRT，因为 SRT 没法控样式和位置。

断行自己算，不依赖渲染库的自动换行。libass 对中文只按字符断不按语义断，
一句话会在词中间折断，观感很差。自己算好断点插换行符才对。
"""

from __future__ import annotations

import unicodedata
from dataclasses import dataclass
from pathlib import Path

# 断在这些标点之后是自然的
_BREAK_AFTER = "，。？！；：、…—"
# 这些不能出现在行首
_NO_LINE_START = "，。？！；：、）】》」』…—·%"
# 这些不能出现在行尾
_NO_LINE_END = "（【《「『"


def display_width(text: str) -> float:
    """按显示宽度算长度。全角算一，半角算半。"""
    total = 0.0
    for ch in text:
        total += 1.0 if unicodedata.east_asian_width(ch) in "WF" else 0.5
    return total


def wrap_chinese(text: str, max_width: float = 15.0, max_lines: int = 2) -> list[str]:
    """中文断行。

    优先在标点后断，其次在宽度上限处断，但要避开标点不能在行首行尾的情况。
    超过行数上限就返回，由调用方决定拆成两条字幕还是截断。
    """
    text = text.strip()
    if not text:
        return []
    if display_width(text) <= max_width:
        return [text]

    lines: list[str] = []
    rest = text
    while rest and len(lines) < max_lines:
        cut = _find_break(rest, max_width)
        if cut <= 0 or cut >= len(rest):
            lines.append(rest)
            rest = ""
            break
        lines.append(rest[:cut].strip())
        rest = rest[cut:].strip()

    if rest:
        # 塞不下的部分并进最后一行，宁可最后一行长一点也不丢字
        lines[-1] = (lines[-1] + rest).strip()
    return [ln for ln in lines if ln]


def _find_break(text: str, max_width: float) -> int:
    """找一个断点。返回切分位置。"""
    width = 0.0
    limit_idx = len(text)
    for i, ch in enumerate(text):
        width += 1.0 if unicodedata.east_asian_width(ch) in "WF" else 0.5
        if width > max_width:
            limit_idx = i
            break

    # 在宽度范围内从后往前找标点
    for i in range(limit_idx - 1, max(0, limit_idx - 8), -1):
        if text[i] in _BREAK_AFTER:
            return i + 1

    # 没有标点就在宽度上限处断，但避开非法位置
    cut = limit_idx
    guard = 0
    while cut > 1 and guard < 6:
        guard += 1
        if cut < len(text) and text[cut] in _NO_LINE_START:
            cut -= 1
            continue
        if text[cut - 1] in _NO_LINE_END:
            cut -= 1
            continue
        break
    return max(1, cut)


@dataclass
class SubtitleCue:
    """一条字幕。"""

    start_s: float
    end_s: float
    text: str
    style: str = "dialogue"

    @property
    def duration_s(self) -> float:
        return self.end_s - self.start_s


def _ass_time(seconds: float) -> str:
    """ASS 的时间格式：时:分:秒.厘秒。"""
    seconds = max(0.0, seconds)
    h = int(seconds // 3600)
    m = int((seconds % 3600) // 60)
    s = seconds % 60
    return f"{h:d}:{m:02d}:{s:05.2f}"


def build_ass(
    cues: list[SubtitleCue],
    width: int = 1080,
    height: int = 1920,
    font: str = "Source Han Sans SC",
    font_size: int = 54,
    max_chars_per_line: int = 15,
    max_lines: int = 2,
    margin_v: int = 180,
) -> str:
    """生成 ASS 字幕文件内容。

    竖屏短剧字幕通常放在下方偏上一点，避开平台的界面元素。
    """
    header = f"""[Script Info]
ScriptType: v4.00+
PlayResX: {width}
PlayResY: {height}
WrapStyle: 2
ScaledBorderAndShadow: yes

[V4+ Styles]
Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding
Style: dialogue,{font},{font_size},&H00FFFFFF,&H000000FF,&H00000000,&H64000000,0,0,0,0,100,100,0,0,1,3,1,2,60,60,{margin_v},1
Style: narration,{font},{int(font_size * 0.92)},&H00E8E8E8,&H000000FF,&H00000000,&H64000000,0,1,0,0,100,100,0,0,1,3,1,2,60,60,{margin_v},1
Style: title,{font},{int(font_size * 1.4)},&H00FFFFFF,&H000000FF,&H00000000,&H96000000,1,0,0,0,100,100,2,0,1,4,2,5,60,60,0,1

[Events]
Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text
"""
    lines = [header]
    # WrapStyle 2 表示只在显式换行符处断行，这正是我们要的：
    # 断点由上面的 wrap_chinese 算好，不让渲染库自作主张。
    for cue in cues:
        if not cue.text.strip():
            continue
        wrapped = wrap_chinese(cue.text, max_chars_per_line, max_lines)
        body = r"\N".join(wrapped)
        style = cue.style if cue.style in ("dialogue", "narration", "title") else "dialogue"
        lines.append(
            f"Dialogue: 0,{_ass_time(cue.start_s)},{_ass_time(cue.end_s)},"
            f"{style},,0,0,0,,{body}"
        )
    return "\n".join(lines) + "\n"


def write_ass(path: str | Path, cues: list[SubtitleCue], **kwargs) -> Path:
    target = Path(path)
    target.parent.mkdir(parents=True, exist_ok=True)
    # ASS 要 UTF-8，加 BOM 能让一些播放器正确识别中文
    target.write_text(build_ass(cues, **kwargs), encoding="utf-8-sig")
    return target


def validate_cues(cues: list[SubtitleCue], max_chars_per_line: int = 15) -> list[str]:
    """字幕自检。装配后闸门要用。"""
    problems: list[str] = []
    for i, cue in enumerate(cues):
        if cue.end_s <= cue.start_s:
            problems.append(f"第 {i + 1} 条字幕时间倒挂")
        if not cue.text.strip():
            problems.append(f"第 {i + 1} 条字幕是空的")
        if cue.duration_s < 0.4:
            problems.append(f"第 {i + 1} 条字幕只显示 {cue.duration_s:.2f} 秒，看不清")
        wrapped = wrap_chinese(cue.text, max_chars_per_line)
        if any(display_width(ln) > max_chars_per_line * 1.6 for ln in wrapped):
            problems.append(f"第 {i + 1} 条字幕断行后仍然超长")
    for a, b in zip(cues, cues[1:]):
        if b.start_s < a.end_s - 0.01:
            problems.append(f"字幕重叠：{a.text[:10]} 与 {b.text[:10]}")
    return problems

File: changji/cpp/src/media/subtitles.cpp
#include "media/subtitles.hpp"

#include <algorithm>
// **MSVC 会顺带把 <cmath> 带进来，GCC 不会。** 这一处 std::fmod 在
// Windows 上编了几个月都没事，第一次在 Linux 上编就挂：
// "'fmod' is not a member of 'std'"。方案里"交叉编译到树莓派"那条要的
// 就是这种东西早点冒出来。
#include <cmath>
#include <cstdio>
#include <fstream>
#include <sstream>

#include "util/paths.hpp"
#include "util/text.hpp"

namespace fs = std::filesystem;

namespace changji::media {

namespace {

#include "media/east_asian_width.inc.hpp"

/// 断在这些标点之后是自然的。
const std::vector<std::string>& break_after() {
    static const std::vector<std::string> v = {"，", "。", "？", "！", "；",
                                               "：", "、", "…", "—"};
    return v;
}

/// 这些不能出现在行首。
const std::vector<std::string>& no_line_start() {
    static const std::vector<std::string> v = {"，", "。", "？", "！", "；",
                                               "：", "、", "）", "】", "》",
                                               "」", "』", "…", "—", "·", "%"};
    return v;
}

/// 这些不能出现在行尾。
const std::vector<std::string>& no_line_end() {
    static const std::vector<std::string> v = {"（", "【", "《", "「", "『"};
    return v;
}

/// 把 UTF-8 串切成一个个字符（**不是字节**）。
///
/// 断行的所有下标都以"第几个字符"计。按字节算的话，一个中文字会被
/// 劈成三段，产出非法 UTF-8——而那个串会进 ASS 文件，播放器直接不显示。
std::vector<std::string> chars_of(const std::string& s) {
    std::vector<std::string> out;
    std::size_t i = 0;
    while (i < s.size()) {
        const std::size_t n =
            text::utf8_char_len(static_cast<unsigned char>(s[i]));
        const std::size_t take = std::min(n, s.size() - i);
        out.push_back(s.substr(i, take));
        i += take;
    }
    return out;
}

char32_t codepoint_of(const std::string& ch) {
    if (ch.empty()) return 0;
    const auto b0 = static_cast<unsigned char>(ch[0]);
    if (b0 < 0x80) return b0;
    if ((b0 & 0xE0) == 0xC0 && ch.size() >= 2) {
        return static_cast<char32_t>(((b0 & 0x1F) << 6) |
                                     (static_cast<unsigned char>(ch[1]) & 0x3F));
    }
    if ((b0 & 0xF0) == 0xE0 && ch.size() >= 3) {
        return static_cast<char32_t>(((b0 & 0x0F) << 12) |
                                     ((static_cast<unsigned char>(ch[1]) & 0x3F) << 6) |
                                     (static_cast<unsigned char>(ch[2]) & 0x3F));
    }
    if ((b0 & 0xF8) == 0xF0 && ch.size() >= 4) {
        return static_cast<char32_t>(((b0 & 0x07) << 18) |
                                     ((static_cast<unsigned char>(ch[1]) & 0x3F) << 12) |
                                     ((static_cast<unsigned char>(ch[2]) & 0x3F) << 6) |
                                     (static_cast<unsigned char>(ch[3]) & 0x3F));
    }
    return 0;
}

bool in(const std::vector<std::string>& set, const std::string& ch) {
    return std::find(set.begin(), set.end(), ch) != set.end();
}

double width_of(const std::string& ch) { return is_wide(codepoint_of(ch)) ? 1.0 : 0.5; }

std::string join(const std::vector<std::string>& v, std::size_t from,
                 std::size_t to) {
    std::string out;
    for (std::size_t i = from; i < to && i < v.size(); ++i) out += v[i];
    return out;
}

/// 找一个断点，返回**第几个字符**处切。
std::size_t find_break(const std::vector<std::string>& chars, double max_width) {
    double width = 0.0;
    std::size_t limit = chars.size();
    for (std::size_t i = 0; i < chars.size(); ++i) {
        width += width_of(chars[i]);
        if (width > max_width) {
            limit = i;
            break;
        }
    }

    // 在宽度范围内从后往前找标点。只往回找 8 个字符：找太远的话，
    // 一行会短得很难看，那还不如在宽度处硬断。
    //
    // **边界要和 Python 的 range(limit-1, max(0, limit-8), -1) 逐个对齐。**
    // 多查一格或少查一格，某些句子的断点就和 Python 差一个字——
    // 那不会报错，只会让两边出的字幕文件不一样。
    if (limit >= 2) {
        const std::size_t stop = limit > 8 ? limit - 8 : 0;
        for (std::size_t i = limit - 1; i > stop; --i) {
            if (in(break_after(), chars[i])) return i + 1;
        }
    }

    // 没有标点就在宽度上限处断，但避开非法位置。
    std::size_t cut = limit;
    int guard = 0;
    while (cut > 1 && guard < 6) {
        ++guard;
        if (cut < chars.size() && in(no_line_start(), chars[cut])) {
            --cut;
            continue;
        }
        if (in(no_line_end(), chars[cut - 1])) {
            --cut;
            continue;
        }
        break;
    }
    return std::max<std::size_t>(1, cut);
}

}  // namespace

bool is_wide(char32_t cp) {
    // 区间表是有序的，二分。121 个区间线性扫也不慢，但断行要对每个字符
    // 调一次，一集几千字就是几十万次。
    std::size_t lo = 0;
    std::size_t hi = sizeof(kWideRanges) / sizeof(kWideRanges[0]);
    while (lo < hi) {
        const std::size_t mid = lo + (hi - lo) / 2;
        if (cp < kWideRanges[mid].lo) {
            hi = mid;
        } else if (cp > kWideRanges[mid].hi) {
            lo = mid + 1;
        } else {
            return true;
        }
    }
    return false;
}

double display_width(const std::string& utf8) {
    double total = 0.0;
    for (const auto& ch : chars_of(utf8)) total += width_of(ch);
    return total;
}

std::vector<std::string> wrap_chinese(const std::string& text, double max_width,
                                      int max_lines) {
    const std::string trimmed = text::strip_ws(text);
    if (trimmed.empty()) return {};
    if (display_width(trimmed) <= max_width) return {trimmed};

    std::vector<std::string> lines;
    std::vector<std::string> rest = chars_of(trimmed);

    while (!rest.empty() && static_cast<int>(lines.size()) < max_lines) {
        const std::size_t cut = find_break(rest, max_width);
        if (cut == 0 || cut >= rest.size()) {
            lines.push_back(text::strip_ws(join(rest, 0, rest.size())));
            rest.clear();
            break;
        }
        lines.push_back(text::strip_ws(join(rest, 0, cut)));
        rest = chars_of(text::strip_ws(join(rest, cut, rest.size())));
    }

    if (!rest.empty()) {
        // 塞不下的部分并进最后一行。**宁可最后一行长一点也不丢字**——
        // 丢字的表现是观众看到一句没说完的话。
        if (lines.empty()) lines.push_back("");
        lines.back() = text::strip_ws(lines.back() + join(rest, 0, rest.size()));
    }

    std::vector<std::string> out;
    for (auto& ln : lines) {
        if (!ln.empty()) out.push_back(ln);
    }
    return out;
}

/// 台词里的字进 Dialogue 行之前要过一遍。
///
/// **花括号是硬伤。** ASS 里 `{` 开始一个特效覆盖块、到 `}` 为止整段被
/// 吞掉。台词里出现一个 `{`，那几个字在成片里就没了——**而且不报错**，
/// 要盯着片子看才发现。剧本是大模型写的，它偶尔会吐出 ASCII 花括号。
/// libass 认 `\\{` 这种写法，渲染成一个字面的大括号。
///
/// **裸换行会把这一行拆断。** Dialogue 是一行一条记录，文本里混进 CR/LF
/// 之后半条记录变成下一行，渲染器多半直接忽略——又是一处静默丢字。
/// 换成空格：真要换行，上面 wrap 出来的那几段之间已经有 \\N 了。
///
/// 反斜杠不动：ASS 没有通用的反斜杠转义，而中文台词里出现裸反斜杠
/// 比出现花括号少得多，动它反而可能改坏本来对的输出。
std::string ass_escape(const std::string& text) {
    std::string out;
    out.reserve(text.size());
    for (const char c : text) {
        if (c == '{' || c == '}') {
            out += '\\';
            out += c;
        } else if (c == '\n' || c == '\r') {
            out += ' ';
        } else {
            out += c;
        }
    }
    return out;
}

std::string ass_time(double seconds) {
    seconds = std::max(0.0, seconds);
    const int h = static_cast<int>(seconds / 3600.0);
    const int m = static_cast<int>(std::fmod(seconds, 3600.0) / 60.0);
    const double s = std::fmod(seconds, 60.0);
    char buf[64];
    std::snprintf(buf, sizeof(buf), "%d:%02d:%05.2f", h, m, s);
    return buf;
}

std::string build_ass(const std::vector<SubtitleCue>& cues,
                      const AssOptions& opt) {
    std::ostringstream os;
    // WrapStyle 2 = 只在显式换行符处断行。断点由上面的 wrap_chinese 算好，
    // 不让渲染库自作主张——libass 对中文只按字符断，会在词中间折断。
    os << "[Script Info]\n"
       << "ScriptType: v4.00+\n"
       << "PlayResX: " << opt.width << "\n"
       << "PlayResY: " << opt.height << "\n"
       << "WrapStyle: 2\n"
       << "ScaledBorderAndShadow: yes\n"
       << "\n"
       << "[V4+ Styles]\n"
       << "Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, "
          "OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, "
          "ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, "
          "Alignment, MarginL, MarginR, MarginV, Encoding\n"
       << "Style: dialogue," << opt.font << "," << opt.font_size
       << ",&H00FFFFFF,&H000000FF,&H00000000,&H64000000,0,0,0,0,100,100,0,0,1,3,"
          "1,2,60,60,"
       << opt.margin_v << ",1\n"
       << "Style: narration," << opt.font << ","
       << static_cast<int>(opt.font_size * 0.92)
       << ",&H00E8E8E8,&H000000FF,&H00000000,&H64000000,0,1,0,0,100,100,0,0,1,3,"
          "1,2,60,60,"
       << opt.margin_v << ",1\n"
       << "Style: title," << opt.font << ","
       << static_cast<int>(opt.font_size * 1.4)
       << ",&H00FFFFFF,&H000000FF,&H00000000,&H96000000,1,0,0,0,100,100,2,0,1,4,"
          "2,5,60,60,0,1\n"
       << "\n"
       << "[Events]\n"
       << "Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, "
          "Effect, Text\n";

    std::string out = os.str();
    for (const auto& cue : cues) {
        if (text::strip_ws(cue.text).empty()) continue;
        const auto wrapped = wrap_chinese(
            cue.text, static_cast<double>(opt.max_chars_per_line), opt.max_lines);
        std::string body;
        for (std::size_t i = 0; i < wrapped.size(); ++i) {
            if (i) body += "\\N";   // ASS 的硬换行
            // **先转义再拼。** 反过来的话会把我们自己刚写的 \\N 也转掉。
            body += ass_escape(wrapped[i]);
        }
        const std::string style =
            (cue.style == "dialogue" || cue.style == "narration" ||
             cue.style == "title")
                ? cue.style
                : "dialogue";
        out += "\nDialogue: 0," + ass_time(cue.start_s) + "," +
               ass_time(cue.end_s) + "," + style + ",,0,0,0,," + body;
    }
    return out + "\n";
}

fs::path write_ass(const fs::path& path, const std::vector<SubtitleCue>& cues,
                   const AssOptions& opt) {
    std::error_code ec;
    fs::create_directories(path.parent_path(), ec);
    std::ofstream f(path, std::ios::binary | std::ios::trunc);
    if (!f) throw std::runtime_error("写不了字幕文件：" + paths::to_utf8(path));
    // UTF-8 BOM。一些播放器靠它才认出中文，没有的话按本地代码页解，
    // 出来的是一屏乱码。
    f << "\xEF\xBB\xBF" << build_ass(cues, opt);
    return path;
}

std::vector<std::string> validate_cues(const std::vector<SubtitleCue>& cues,
                                       int max_chars_per_line) {
    std::vector<std::string> problems;
    for (std::size_t i = 0; i < cues.size(); ++i) {
        const auto& cue = cues[i];
        const std::string n = std::to_string(i + 1);
        if (cue.end_s <= cue.start_s) {
            problems.push_back("第 " + n + " 条字幕时间倒挂");
        }
        if (text::strip_ws(cue.text).empty()) {
            problems.push_back("第 " + n + " 条字幕是空的");
        }
        if (cue.duration_s() < 0.4) {
            char buf[32];
            std::snprintf(buf, sizeof(buf), "%.2f", cue.duration_s());
            problems.push_back("第 " + n + " 条字幕只显示 " + buf + " 秒，看不清");
        }
        const auto wrapped =
            wrap_chinese(cue.text, static_cast<double>(max_chars_per_line));
        for (const auto& ln : wrapped) {
            if (display_width(ln) > max_chars_per_line * 1.6) {
                problems.push_back("第 " + n + " 条字幕断行后仍然超长");
                break;
            }
        }
    }
    for (std::size_t i = 0; i + 1 < cues.size(); ++i) {
        const auto& a = cues[i];
        const auto& b = cues[i + 1];
        if (b.start_s < a.end_s - 0.01) {
            problems.push_back("字幕重叠：" + text::truncate_utf8(a.text, 10) +
                               " 与 " + text::truncate_utf8(b.text, 10));
        }
    }
    return problems;
}

}  // namespace changji::media
